### PR TITLE
8345936: Call ClassLoader.getResourceAsByteArray only for multi-release jar

### PR DESCRIPTION
**Pull request by calvinccheung**
This patch stores whether a jar is multi-release in the SharedClassPathEntry during CDS dump time.
At runtime, it only calls ClassLoader.getResourceAsByteArray if the jar is a multi-release jar.
As a result, the number of java calls is reduced.

Passed tiers 1, 3, 4 testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345936](https://bugs.openjdk.org/browse/JDK-8345936): Call ClassLoader.getResourceAsByteArray only for multi-release jar (**Enhancement** - P4)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22696/head:pull/22696` \
`$ git checkout pull/22696`

Update a local copy of the PR: \
`$ git checkout pull/22696` \
`$ git pull https://git.openjdk.org/jdk.git pull/22696/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22696`

View PR using the GUI difftool: \
`$ git pr show -t 22696`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22696.diff">https://git.openjdk.org/jdk/pull/22696.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22696#issuecomment-2537465441)
</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoader.cpp [hunk 1]
```diff
@@ -852,7 +854,10 @@ bool ClassLoader::add_to_app_classpath_entries(JavaThread* current,
   }
 
   if (entry->is_jar_file()) {
-    ClassLoaderExt::process_jar_manifest(current, entry);
+    bool is_multi_release = ClassLoaderExt::process_jar_manifest(current, entry);
+    if (is_multi_release) {
+      entry->set_multi_release_jar();
```
Comment (iklam) [inline src/hotspot/share/classfile/classLoader.cpp:859]: Maybe the set_multi_release_jar() call should be moved into process_jar_manifest()? That way, the logic doesn't need to be split in two places.

**Review by iklam — CHANGES_REQUESTED** [review]
Looks good overall. I have one suggestion for simplification.

File: src/hotspot/share/classfile/classLoader.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/classfile/classLoader.cpp]: That's a good idea. I've pushed an update.

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable.

Thanks

**Review by iklam — APPROVED** [review]
LGTM

**Comment by calvinccheung** [comment]
Thanks @iklam and @dholmes-ora for the review.

/integrate

**Pull request closed without merging**